File: Java/src/main/java/SimUDuck/Duck.java
package SimUDuck;

public abstract class Duck {
    public String duckType;
    FlyBehaviour flyBehaviour;
    QuackBehaviour quackBehaviour;

    public void performFly(){
        if(flyBehaviour==null){
            System.out.println("This duck can't fly");
        }else {
            flyBehaviour.fly();
        }
    }

    public void performQuack(){
        if(quackBehaviour==null){
            System.out.println("This Duck can't Quack");
        }else{
            quackBehaviour.quack();
        }
    }

    public void display(){
        System.out.println("This is a Duck of type:  "+ duckType);
    }

    public void setFlyBehaviour(FlyBehaviour flyBehaviour){
        this.flyBehaviour = flyBehaviour;
    }

    public void setQuackBehaviour(QuackBehaviour quackBehaviour){
        this.quackBehaviour = quackBehaviour;
    }
}

File: TS/Intro-Basics/Duck.ts
export class Duck{  
    public species:string; 
    constructor(species:string){ 
        this.species = species;
    } 

    flyable : Flyable; 

    quackable : Quakable; 

    public performQuack(){  

        this.quackable.Quack(); 
    } 

    public performFly(){ 
        this.flyable.fly(); 
    }

    public swim(){ 

    } 

    public display(){  
        console.log("Looks like a ",this.species);
    }  

    public flyBehaviourSetter( flyBehaviour : Flyable){ 
        this.flyable = flyBehaviour; 
    } 

    public quackBehaviourSetter(quackBehaviour : Quakable){ 
        this.quackable = quackBehaviour;
    } 

} 

export interface Flyable{  
        fly() : void;   

} 

export interface Quakable{ 
        Quack():void; 
} 

